client/callout.go

```diff
@@ -0,0 +1,153 @@
+package client
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	defaultTimeout             = time.Minute
+	defaultDialTimeout         = 5 * time.Second
+	defaultTLSHandshakeTimeout = 5 * time.Second
+)
+
+type Caller interface {
+	Get(url string, options ...RequestOption) ([]byte, error)
+	Head(url string, options ...RequestOption) ([]byte, error)
+	Post(url, body string, options ...RequestOption) ([]byte, error)
+}
+
+type Callout struct {
+	defaultHeaders map[string]string
+	defaultTimeout time.Duration
+	defaultRetries int
+	skipTLSVerify  bool
+}
+
+// Ensure Callout implements Caller interface
+var _ Caller = &Callout{}
+
+func New(options ...CalloutOption) *Callout {
+	callout := &Callout{
+		defaultTimeout: defaultTimeout,
+	}
+
+	for _, option := range options {
+		option(callout)
+	}
+
+	return callout
+}
+
+func (c *Callout) Get(url string, options ...RequestOption) ([]byte, error) {
```

Comment:
So the interface is:
1. On success, we get a `[]byte` of the response.
1. On success, On `jsonValue`, we get `[]byte` of response, And the response parsed for you.
1. On success, On `writeBody`, we **dont** get `[]byte` of response, And do get the response copied for you.
1. On failure, we get `ResponseError` with `[]byte` of the response in the error.
1. (_edge case, not validated_) On failure, On `writeBody`, we `ResponseError` **without** `[]byte` of the response in the error, And do get the response copied for you.

🤔 Just seems like different behavior around responses, no? I can't think of an elegant way to do it, but It'd be nice if at least `1.` and `2.` and `3.` behaved the same. I wouldn't hold the PR on something like this though.

Reply:
This is a tricky one that @TheBoysAreZachInTown and I talked about a bunch.

The use-case I had in mind for `writeBody` is a response that you want to write directly to a file, either because you don't need access to it or it's too big to fit in memory. So I didn't return it because of that second case. It wouldn't be difficult to write the body to the writer and also return it, but that could potentially run out of memory. Maybe it would be more consistent if cases `2` and `3` both **don't** return the body?

There is certainly the strange edge-case where `writeBody` is used and the status code is 4XX (or 5XX after all the retries). The response body is written to the writer and is not set on `ResponseError.Body`. I can't think of a good way around that due to the memory issue above.

Reply:
Thanks for the explanation! Really enjoy knowing the thought process. 👍🏾